nexosim/examples/bench.rs

```diff
@@ -0,0 +1,221 @@
+//!                    brew time   └───►│            │ command  │            │   │
+//! Brew time dial ●───────────────────►│ Controller ├─────────►│ Water pump ├───┘
+//!                      (>0)      ┌───►│            │ (on|off) │            │
+//!                                │    └────────────┘          └────────────┘
+//!                    trigger     │
+//!   Brew command ●───────────────┘
+//!                      (-)
+//! ```
+
+use std::time::Duration;
+
+use nexosim::ports::{EventQueue, EventSlot, EventSource, QuerySource};
+use nexosim::simulation::{Mailbox, SimInit, SimulationError, SourceId};
+use nexosim::time::MonotonicTime;
+
+mod espresso_machine;
+
+pub use espresso_machine::{Controller, Pump, Tank};
+
+pub fn bench((pump_flow_rate, init_tank_volume): (f64, f64)) -> SimInit {
+    // Models.
+    let mut pump = Pump::new(pump_flow_rate);
+    let mut controller = Controller::new();
+    let mut tank = Tank::new(init_tank_volume);
+
+    // Mailboxes.
+    let pump_mbox = Mailbox::new();
+    let controller_mbox = Mailbox::new();
+    let tank_mbox = Mailbox::new();
+
+    // Connections.
+    controller.pump_cmd.connect(Pump::command, &pump_mbox);
+    tank.water_sense
+        .connect(Controller::water_sense, &controller_mbox);
+    pump.flow_rate.connect(Tank::set_flow_rate, &tank_mbox);
+
+    // Sinks.
+
+    // Controller.
+    let pump_cmd = EventQueue::new();
+    controller.pump_cmd.connect_sink(&pump_cmd);
+
+    // Pump.
+    let flow_rate = EventSlot::new();
+    pump.flow_rate.connect_sink(&flow_rate);
+
+    // Tank.
+    let water_sense = EventSlot::new();
+    tank.water_sense.connect_sink(&water_sense);
+
+    // Sources.
+
+    // Controller.
+    let mut brew_time = EventSource::new();
+    brew_time.connect(Controller::brew_time, &controller_mbox);
+
+    let mut brew_cmd = EventSource::new();
+    brew_cmd.connect(Controller::brew_cmd, &controller_mbox);
+
+    // Tank.
+    let mut fill = EventSource::new();
+    fill.connect(Tank::fill, &tank_mbox);
+
+    let mut volume = QuerySource::new();
+    volume.connect(Tank::volume, &tank_mbox);
+
+    // Bench assembly.
+    let mut bench = SimInit::new()
+        .add_model(controller, controller_mbox, "controller")
+        .add_model(pump, pump_mbox, "pump")
+        .add_model(tank, tank_mbox, "tank");
+
+    bench
+        .add_event_sink_endpoint(pump_cmd.into_reader(), "pump_cmd")
+        .unwrap();
+    bench
+        .add_event_sink_endpoint(flow_rate, "flow_rate")
+        .unwrap();
+    bench
+        .add_event_sink_endpoint(water_sense, "water_sense")
+        .unwrap();
+
+    bench
+        .add_event_source_endpoint(brew_time, "brew_time")
+        .unwrap();
+    bench
+        .add_event_source_endpoint(brew_cmd, "brew_cmd")
+        .unwrap();
+    bench.add_event_source_endpoint(fill, "fill").unwrap();
+
+    bench.add_query_source_endpoint(volume, "volume").unwrap();
```

Comment:
These unwrap everywhere are a bit painful and go against the builder pattern.

My suggestion is to design these methods so they follow the builder pattern an can be chained with a `?.`, making the bench return a `Result<SimInit, BenchError>`.